src/components/oauth/OAuthCodeHandler.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * @author sarahr
+ */
+
+import React, { useCallback, useEffect, useState } from "react";
+import { useRouter } from "next/router";
+import { useQuery } from "react-query";
+import { build } from "@cyverse-de/ui-lib";
+
+import {
+    APP_CATEGORIES_QUERY_KEY,
+    getPrivateCategories,
+} from "serviceFacades/apps";
+import {
+    OAUTH_CALLBACK_QUERY_KEY,
+    doOAuthCallback,
+} from "serviceFacades/oauth";
+
+import OAuthErrorHandler from "./OAuthErrorHandler";
+import OAuthLinkingMessage from "./OAuthLinkingMessage";
+import ids from "./ids";
+import { getListingPath } from "components/apps/utils";
+import systemId from "components/models/systemId";
+import { useUserProfile } from "contexts/userProfile";
+
+const ERR_SERVICE = "general_service_error";
+
+function OAuthCodeHandler(props) {
+    const { baseId, apiName, code, stateId } = props;
+    const handlerId = build(baseId, ids.OAUTH_CODE_HANDLER);
+    const router = useRouter();
+    const [userProfile] = useUserProfile();
+    const [categoryQueryKey, setCategoryQueryKey] = useState([
+        APP_CATEGORIES_QUERY_KEY,
+        userProfile?.id,
+    ]);
+    const [listingUrl, setListingUrl] = useState(
+        getListingPath(null, null, null, null, null, null, false)
+    );
+
+    const determineListingUrl = useCallback(
+        (data) => {
+            const hpcCategory = data.categories.find(
+                (cat) => cat.system_id === systemId.agave
+            );
+            setListingUrl(
+                getListingPath(
+                    null,
+                    null,
+                    null,
+                    null,
+                    null,
+                    JSON.stringify(hpcCategory),
+                    false
+                )
```

Comment:
Redirecting users here means that this component is specific to Tapis If we ever add support for other third-party APIs, we'll have to revisit this.